linera-witty/src/wit_generation.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Generation of WIT files.
+
+use std::collections::BTreeMap;
+
+use genawaiter::{rc::gen, yield_};
+
+pub use crate::type_traits::RegisterWitTypes;
+
+/// Generates WIT snippets for an interface.
+pub trait WitInterface {
+    /// The [`WitType`][`crate::WitType`]s that this interface uses.
+    type Dependencies: RegisterWitTypes;
+
+    /// The name of the package the interface belongs to.
+    fn wit_package() -> &'static str;
+
+    /// The name of the interface.
+    fn wit_name() -> &'static str;
+
+    /// The WIT definitions of each function in this interface.
+    fn wit_functions() -> Vec<String>;
+}
+
+/// Helper type to write a [`WitInterface`] to a file.
+#[derive(Clone, Debug)]
+pub struct WitInterfaceWriter {
+    package: &'static str,
+    name: &'static str,
+    types: BTreeMap<String, String>,
+    functions: Vec<String>,
+}
+
+impl WitInterfaceWriter {
+    /// Prepares a new [`WitInterfaceWriter`] to write the provided `Interface`.
+    pub fn new<Interface>() -> Self
+    where
+        Interface: WitInterface,
+    {
+        let mut types = BTreeMap::new();
+
+        Interface::Dependencies::register_wit_types(&mut types);
+
+        WitInterfaceWriter {
+            package: Interface::wit_package(),
+            name: Interface::wit_name(),
+            types,
+            functions: Interface::wit_functions(),
+        }
+    }
+
+    /// Returns an [`Iterator`] with the file contents of the WIT interface file.
+    pub fn generate_file_contents(&self) -> impl Iterator<Item = &str> {
+        gen!({
```

Comment:
Can we stick to idiomatic Rust?

Here is an example of code generator (with indentation!) that is super basic yet completely fine:
https://github.com/zefchain/serde-reflection/blob/main/serde-generate/src/cpp.rs#L71